File: 2017/rust/src/bin/day04.rs
use std::io;
use std::io::BufRead;
use std::collections::HashSet;

fn main() {
    let mut lines: Vec<Vec<String>> = vec![];
    for line in io::stdin().lock().lines() {
        let line = line.unwrap()
            .split_whitespace()
            .map(str::to_owned)
            .collect();
        lines.push(line);
    }

    println!("Part 1: {}", count_valid(&lines, is_valid));
    println!("Part 1: {}", count_valid(&lines, is_valid_anagram));
}

fn count_valid<P>(lines: &Vec<Vec<String>>, predicate: P) -> usize
where P: FnMut(&&Vec<String>) -> bool {
    lines.iter().filter(predicate).count()
}

fn is_valid(line: &&Vec<String>) -> bool {
    let s: HashSet<&String> = line.iter().collect();
    s.len() == line.len()
}

fn is_valid_anagram(line: &&Vec<String>) -> bool {
    let s: HashSet<String> = line.iter()
        .map(|word| {
            let mut chars: Vec<char> = word.chars().collect();
            chars.sort();
            chars.into_iter().collect()
        })
    .collect();
    s.len() == line.len()
}

File: 2022/ruby/day04.rb
require 'scanf'

data = <<~EXAMPLE
2-4,6-8
2-3,4-5
5-7,7-9
2-8,3-7
6-6,4-6
2-6,4-8
EXAMPLE

data = $stdin.read

lines = data.lines(chomp: true).map { |l| l.scanf("%d-%d,%d-%d") }

p1 = lines.count do |a, b, c, d|
  a >= c && b <= d || c >= a && d <= b
end

p2 = lines.count do |a, b, c, d|
  a >= c && a <= d ||
  b >= c && b <= d ||
  c >= a && c <= b ||
  d >= a && d <= b
end

puts "Part 1", p1
puts "Part 2", p2
